### DISCUSSION TITLE
Consistency Issue?

### DISCUSSION THREAD

**Discussion by erich-team** [Dynamoid/dynamoid#493] [Q&A]
I'm not entirely sure this is an issue, with either dynamoid or dynamodb or just just an issue with how I'm approaching this.  So, I'm posting this here instead of as an issue.

### The Background

Given the models

    class ParentClass
      include SubmissionPackage::Builder
      table
      field :url, :string
      field :full_response, :map, default: {}
    end

    class ChildClass
      def response=(data)
        self.url=data[:secure_url]
        super(data)
      end

       def call_external_system
         #calls some external system with a callback url supplied by us
       end

       after_create :call_external_system
      def enqueue_evaluate_response
          #enqueue a job to do something the the response and URL
      end
    end

I instantiate `ChildClass` which kicks off what is likely the familiar process of calling an external system to do some work and call back in our system with the response. (`ChildClass#call_external_system`). 

    class ResponsesController
      def create
        child_object = ChildClass.find(params[:id])
        child_object.full_response =  params[:response]
        child_object.save
        child_object.enqueue_evaluate_response
      end
    end

`ChildClass#full_response` is called,  `#url` is set, `#full_response` is set, the object is persisted and `#enqueue_evaluate_response` is called, kicking off a job where something is done with URL and the response.  

    class SomeJob<ApplicationJob
      def perform(child_object)
        puts child_object.url
      end
    end

This is a pattern I've used hundreds of times in this application.

The callback into the system usually takes place on the order minutes after ChildClass is initially instantiated.  

### The Complication

Put simply `#url` and `#full_response` are always `nil` inside `SomeJob`, even, if `SomeJob` later.  Not surprisingly the values in `#attributes` and `#attributes_before_type_case` are also nil.  I have checked the AWS console and URL is populated.

#### To further complicate matters:
    SomeClass.find('asdf', consistent_read: true).attributes_before_type_cast[:full_response] # NOT NILL
    SomeClass.find('asdf', consistent_read: true).attributes_before_type_cast[:url] # NIL

#### Similarly

    SomeClass.find('asdf', consistent_read: true).attributes[:full_response] # NOT NILL
    SomeClass.find('asdf', consistent_read: true).attributes[:url] # NIL

#### Lastly

    SomeClass.consistent.find(:id=>'asdf').first.attributes[:full_response] # NIL
    SomeClass.consistent.find(:id=>'asdf').first.attributes[:url] # NIL
    SomeClass.consistent.find(:id=>'asdf').first.attributes_before_type_cast[:full_response] # NIL
    SomeClass.consistent.find(:id=>'asdf').first.attributes_before_type_cast[:url] # NIL

    
### The Conclusion

I'm crazy.  Or I'm confused.  Or there's a bug.  Keep in mind, the DynamoDB console shows all the data being there.

**Comment by andrykonchin** [comment]
Could you please provide some minimal but still complete code examples to reproduce the issue? Right now it isn't clear what is an expected result in your examples.

**Reply by erich-team** [reply]
In trying to (a) obscure business logic of our application and (b) simplify the examples, I have probably overshot which is made worse by being an entirely asynchronous affair.  

In essence, hours after the last create or update, reading a model from dynamodb still returns nil for `#url`field, even when forcing consistent read.

But let me see if I can put together something with a bit more meat.

**Reply by andrykonchin** [reply]
As far as I understand in your example some attribute (`url`) is present in an item in a DynamoDB table but somehow after loading Dynamoid model this field is `nil`. Am I right?

So could you please provide an example to reproduce the issue:
- item from a DynamoDB table
- and model class declaration

If `url` field is declared in a model (with `field` declaration) and item in a table has this attribute - it should be available in a model as well. If it isn't - it's definitely a bug.
